data_model/src/domain.rs

```diff
@@ -30,91 +30,252 @@ pub struct GenesisDomain {
             ))
             .collect(),
             asset_definitions: btree_map::BTreeMap::default(),
-            metadata: Metadata::new(),
+            metadata: Metadata::default(),
             logo: None,
         }
     }
 }
 
-/// Named group of [`Account`] and [`Asset`](`crate::asset::Asset`) entities.
+/// Builder which can be submitted in a transaction to create a new [`Domain`]
 #[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
-pub struct Domain {
-    /// Identification of this [`Domain`].
-    pub id: Id,
-    /// Accounts of the domain.
-    pub accounts: AccountsMap,
-    /// Assets of the domain.
-    pub asset_definitions: AssetDefinitionsMap,
-    /// Metadata of this domain as a key-value store.
-    pub metadata: Metadata,
-    /// IPFS link to domain logo
-    pub logo: Option<IpfsPath>,
+pub struct NewDomain {
+    id: <Domain as Identifiable>::Id,
+    logo: Option<IpfsPath>,
+    metadata: Metadata,
 }
 
-impl PartialOrd for Domain {
+impl PartialOrd for NewDomain {
     #[inline]
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.id.cmp(&other.id))
     }
 }
 
-impl Ord for Domain {
+impl Ord for NewDomain {
     #[inline]
     fn cmp(&self, other: &Self) -> Ordering {
         self.id.cmp(&other.id)
     }
 }
 
-impl Domain {
-    /// Construct [`Domain`] from [`Id`].
-    pub fn new(id: Id) -> Self {
+impl NewDomain {
+    #[must_use]
+    fn new(id: <Domain as Identifiable>::Id) -> Self {
         Self {
             id,
-            accounts: AccountsMap::new(),
-            asset_definitions: AssetDefinitionsMap::new(),
-            metadata: Metadata::new(),
             logo: None,
+            metadata: Metadata::default(),
         }
     }
 
-    /// Domain constructor with pre-setup accounts. Useful for testing purposes.
-    pub fn with_accounts(name: &str, accounts: impl IntoIterator<Item = Account>) -> Self {
-        let accounts_map = accounts
-            .into_iter()
-            .map(|account| (account.id.clone(), account))
-            .collect();
-        #[allow(clippy::expect_used)]
-        Self {
-            id: Id::new(name).expect("must be valid"),
-            accounts: accounts_map,
-            asset_definitions: AssetDefinitionsMap::new(),
-            metadata: Metadata::new(),
-            logo: None,
+    /// Add [`logo`](IpfsPath) to the domain replacing previously defined value
+    #[must_use]
+    pub fn with_logo(mut self, logo: IpfsPath) -> Self {
+        self.logo = Some(logo);
+        self
+    }
+
+    /// Add [`Metadata`] to the domain replacing previously defined value
+    #[must_use]
+    pub fn with_metadata(mut self, metadata: Metadata) -> Self {
+        self.metadata = metadata;
+        self
+    }
+
+    /// Construct [`Domain`]
+    #[must_use]
+    #[cfg(feature = "mutable_api")]
+    pub fn build(self) -> Domain {
```

Comment:
So now we can't create new domains without `mutable_api` feature? Then whole `NewDomain` implementation has no meaning without that method, all impl block should be marked as `mutable_api`

Reply:
yes, it's not possible to construct `Domain` without `mutable_api` feature which is intended to be used only by `iroha_core`. `NewDomain` is to be constructed by the client and submitted in a txn to `iroha_core`

Reply:
@Arjentix @appetrosyan I'd like to close this, if there aren't any suggestions. I think this is fine